cardano-node-chairman/src/Chairman/IO/Network/Socket.hs

```diff
@@ -17,42 +11,32 @@ module Chairman.IO.Network.Socket
 import           Control.Exception (IOException, handle)
 import           Control.Monad
 import           Data.Bool
-import           Data.Either
-import           Data.Eq
 import           Data.Function
 import           Data.Functor
 import           Data.Int
 import           Data.Maybe
-import           Foreign.C.Error (Errno (..), eCONNREFUSED)
 import           Network.Socket (Family (AF_INET), SockAddr (..), Socket, SocketType (Stream))
 import           System.IO (FilePath, IO)
 import           Text.Show (show)
 
-import qualified Data.List as L
-import qualified GHC.IO.Exception as IO
 import qualified Network.Socket as IO
 import qualified System.Directory as IO
 import qualified UnliftIO.Exception as IO
 
 -- | Check if a TCP port is open
 isPortOpen :: Int -> IO Bool
-isPortOpen port = handle (\(_ :: IOException) -> return False) $ do
+isPortOpen port = do
   socketAddressInfos <- IO.getAddrInfo Nothing (Just "127.0.0.1") (Just (show port))
   case socketAddressInfos of
-    socketAddressInfo:_ -> canConnect (IO.addrAddress socketAddressInfo)
+    socketAddressInfo:_ ->
+      handle (return . const @Bool @IOException False) $
+        canConnect (IO.addrAddress socketAddressInfo) $> True
     [] -> return False
 
 -- | Check if it is possible to connect to a socket address
-canConnect :: SockAddr -> IO Bool
+canConnect :: SockAddr -> IO ()
 canConnect sockAddr = IO.bracket (IO.socket AF_INET Stream 6) IO.close' $ \sock -> do
-  result <- IO.try $ IO.connect sock sockAddr
-  case result of
-    Right () -> return True
-    Left e
-      | (Errno <$> IO.ioe_errno e) == Just eCONNREFUSED -> return False
-      | "WSAECONNREFUSED" `L.isInfixOf` show e -> return False
-      | "WSAECONNRESET" `L.isInfixOf` show e -> return False
-      | otherwise -> IO.throwIO e
```

Comment:
Exception has been moved to the caller of `canConnect`.  This means exception handling can happen in one place.

Moreover, if an exception is thrown at all, we'll report that the port was not open.  This is to avoid tests being flaky, which has been the case when some other unhandled variant of `IOException` was thrown.